actor/todo-sql/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+};
+use wasmcloud_interface_logging::{debug, info, warn};
+use once_cell::sync::Lazy;
+use minicbor::{decode,Decode};
+
+type Db = SqlDbSender<WasmHost>;
+const TABLE_NAME :&str = "todos";
+/// limit characters to avoid sql injection attacks
+/// - english letters uppercase & lowercase
+/// - digits 0-9
+/// - space, comma, period, underscore, dash, slash
+/// that's it!
+static REGEX: Lazy<regex::Regex> = Lazy::new(|| {
+    let re = regex::Regex::new(r"^[a-zA-Z0-9 ,._-/]+$").unwrap();
+    re
+});
+
+#[derive(Serialize, Deserialize)]
+struct InputTodo {
+    title: String,
+    order: Option<i32>,
+}
+#[derive(Serialize, Deserialize)]
+struct UpdateTodo {
+    title: Option<String>,
+    completed: Option<bool>,
+    order: Option<i32>,
+}
+// because 'order' is a keyword in sql, we've renamed the field 'priority',
+// but for serde serialization, leave it as 'order' so that it works
+// with existing javascript written for this app
+#[derive(Serialize, Deserialize, Decode)]
+struct Todo {
+    #[n(0)] url: String,
+    #[n(1)] title: String,
+    #[n(2)] completed: bool,
+    #[serde(rename="order")]
+    #[n(3)] priority: i32,
+}
+
+impl Todo {
+    fn new(url: String, title: String, priority: i32) -> Self {
+        Self {
+            url,
+            title,
+            completed: false,
+            priority,
+        }
+    }
+
+    fn update(self, update: UpdateTodo) -> Todo {
+        Todo {
+            url: self.url,
+            title: update.title.unwrap_or(self.title),
+            completed: update.completed.unwrap_or(self.completed),
+            priority: update.order.unwrap_or(self.priority),
+        }
+    }
+}
+
+// check safety of any user input string before it used in an sql statement
+// If the provider (&interface) supported prepared statements,
+// this could allow more characters
+fn check_safety(tag: &str, uncertain_input: &str) -> Result<(), SqlDbError> {
+    if !REGEX.is_match(&uncertain_input) {
+        return Err(SqlDbError::new("invalid",
+           format!("{} contains invalid characters", tag)))
+    }
+    Ok(())
+}
+
+/// create an empty table with the proper schema
+async fn create_table(ctx: &Context, client: &Db) -> Result<(),SqlDbError> {
+
+    let sql = format!(r#"create table {} (
+            id varchar(36) not null,
+            url varchar(42) not null,
+            title varchar(100) not null,
+            priority int4 not null default 0,
+            completed bool not null default false
+        );"#, TABLE_NAME);
+    let _resp = client.execute(&ctx, &sql) .await?;
+    Ok(())
+}
+
+/// insert a new to-do item
+async fn create_todo(ctx: &Context, client: &Db, input: InputTodo) -> Result<Todo,SqlDbError> {
+
+    check_safety("title", &input.title)?;
+    let id = generate_guid().await.unwrap();
+    let todo = Todo::new(
+        format!("/api/{}", id),
+        input.title,
+        input.order.unwrap_or(0),
+    );
+
+    let _resp = client
+        .execute(&ctx, &format!(r#"
```

Comment:
If we use named parameters in the text string, then `execute` could take a string for the SQL statement and a `parameter map` for the arguments, which would implicitly take care of SQL injection and sanitization. Right now, someone could use a cleverly-placed apostrophe to create a SQL injection attack here. 

We don't want developers to have to worry about that kind of stuff by default, "secure by default" should hopefully apply to our SQL db interface, too.

Reply:
e.g.
```
client
    .execute(&ctx, "insert into mytable (id, url, title, order, completed) values ({id}, {title}, {order}, {completed})",
         params_map).await?;
```

In this sample, it's implied that we (sqldb provider) will take care of apostrophes (or whatever delimiter the DB requires) as well as per-field sanitization and escaping.

The value type of the parameter map could be a simple enum with variants like `String` and `Number` , which would allow us heuristics to determine how the final query statement should be generated.

The goal as always is to remove as much boilerplate/busywork from developers as possible, and using the `format` macro to build _safe_ SQL statements is busywork.

Reply:
will add parameterized statements in a subsequent version